chore: Cleanup incorrect links in CHANGELOG.md

SuperSodaSea opened the pull request:
<!-- Thanks for submitting a pull request! Please provide enough information so that others can review your pull request. The two fields below are mandatory. -->

<!-- Please remember to update CHANGELOG.md at the root of the project if you have not done so. -->

## Summary

<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->

I happened to find some of the PRs' links in CHANGELOG.md are incorrect (different IDs in text and URLs), due to mistake when copying and pasting them:

https://github.com/facebook/jest/blob/48ddbf49535290aa9cd0e990f44d583c1baab96e/CHANGELOG.md?plain=1#L862

So I wrote a small script like this to find all of them (hopefully):

```js
const fs = require('fs');

const regex =
  /\[#(\d+)\]\(https:\/\/github.com\/facebook\/jest\/pull\/(\d+)\)/g;
const data = fs.readFileSync('CHANGELOG.md', 'utf-8');
for (const match of data.matchAll(regex))
  if (match[1] !== match[2]) console.error(`Error: ${match[0]}`);
```

And it outputs:

```
Error: [#12763](https://github.com/facebook/jest/pull/12899)
Error: [#11523](https://github.com/facebook/jest/pull/11567)
Error: [#8751](https://github.com/facebook/jest/pull/8829)
Error: [#10393](https://github.com/facebook/jest/pull/10400)
Error: [#10217](https://github.com/facebook/jest/pull/10237)
Error: [#8448](https://github.com/facebook/jest/pull/8665)
Error: [#7985](https://github.com/facebook/jest/pull/8003)
Error: [#7809](https://github.com/facebook/jest/pull/7972)
Error: [#6792](https://github.com/facebook/jest/pull/6810)
Error: [#5948](https://github.com/facebook/jest/pull/5953)
Error: [#5858](https://github.com/facebook/jest/pull/3497)
Error: [#5154](https://github.com/facebook/jest/pull/5502)
Error: [#4837](https://github.com/facebook/jest/pull/4629)
```

I manually checked them one by one and fixed them in this PR.

## Test plan

<!-- Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots / videos if the pull request changes UI. -->

~(Maybe we can add a step somewhere in the CI to detect such error.)~

A CI step is added to check this, see discussion below.

SimenB commented:
Ah nice, thanks @SuperSodaSea!

>So I wrote a small script like this to find all of them (hopefully):

Can you add that script to the repo? We can run it as part of some CI run to verify links are correct

SuperSodaSea commented:
@SimenB I added `scripts/checkChangelog.mjs`, it will emit error like this when error is detected:

```
CHANGELOG.md:298:72: error: Link is incorrect: [#12763](https://github.com/facebook/jest/pull/12899)
```

It can be run via `yarn check-changelog`, and I added a `check-changelog` step to the `Lint` job of CI.

SimenB commented on scripts/checkChangelog.mjs:
```diff
@@ -0,0 +1,31 @@
+  ++lineNumber;
+}
+
+process.exit(+error);
```
```suggestion
process.exit(error ? 1 : 0);
```

SimenB reviewed (approved):
thanks!
